diff --git a/content/browser/browser_context.cc b/content/browser/browser_context.cc
index 13e78598fea92df8a30475118a3cc2cd48f2a817..3c0abb011cbafd48de1945d6222be2ada2ab14e0 100644
--- a/content/browser/browser_context.cc
+++ b/content/browser/browser_context.cc
@@ -50,6 +50,7 @@
 #include "content/public/browser/browser_thread.h"
 #include "content/public/browser/content_browser_client.h"
 #include "content/public/browser/render_process_host.h"
+#include "content/public/browser/render_view_host.h"
 #include "content/public/browser/shared_cors_origin_access_list.h"
 #include "content/public/browser/site_instance.h"
 #include "content/public/browser/storage_partition_config.h"
@@ -626,4 +627,10 @@ variations::VariationsClient* BrowserContext::GetVariationsClient() {
   return nullptr;
 }
 
+// static
+StoragePartitionImplMap* BrowserContext::GetStoragePartitionImplMap() {
+  return static_cast<StoragePartitionImplMap*>(
+      GetUserData(kStoragePartitionMapKeyName));
+}
+
 }  // namespace content
