diff --git a/content/public/browser/browser_context.h b/content/public/browser/browser_context.h
index 429b7d0188e7c2efc40df0fb5cb3d8006a2368a9..97bd041fe2d3b43a9fed3db45773c64b895b1bc3 100644
--- a/content/public/browser/browser_context.h
+++ b/content/public/browser/browser_context.h
@@ -93,6 +93,7 @@ class SiteInstance;
 class StorageNotificationService;
 class StoragePartition;
 class StoragePartitionConfig;
+class StoragePartitionImplMap;
 class SSLHostStateDelegate;
 
 // A mapping from the scheme name to the protocol handler that services its
@@ -132,6 +133,8 @@ class CONTENT_EXPORT BrowserContext : public base::SupportsUserData {
                                                SiteInstance* site_instance,
                                                bool can_create = true);
 
+  StoragePartitionImplMap* GetStoragePartitionImplMap();
+
   // Returns a StoragePartition for the given StoragePartitionConfig. By
   // default this will create a new StoragePartition if it doesn't exist,
   // unless |can_create| is false.
