diff --git a/content/browser/browser_interface_binders.cc b/content/browser/browser_interface_binders.cc
index 75838b4f1a0c3f006125a97d91dce15d6831359b..e2f49201d3cdca0e071f19a4227e463ff9dae101 100644
--- a/content/browser/browser_interface_binders.cc
+++ b/content/browser/browser_interface_binders.cc
@@ -510,6 +510,9 @@ void PopulateFrameBinders(RenderFrameHostImpl* host, mojo::BinderMap* map) {
   map->Add<blink::mojom::FileSystemManager>(base::BindRepeating(
       &RenderFrameHostImpl::GetFileSystemManager, base::Unretained(host)));
 
+  map->Add<blink::mojom::DomStorageProvider>(base::BindRepeating(
+      &RenderFrameHostImpl::GetDomStorageProvider, base::Unretained(host)));
+
   map->Add<blink::mojom::GeolocationService>(base::BindRepeating(
       &RenderFrameHostImpl::GetGeolocationService, base::Unretained(host)));
 
