diff --git a/extensions/browser/BUILD.gn b/extensions/browser/BUILD.gn
index 86076a326933d014272a35c2bdb04bb8e3bcb746..c7e6ce7637918fccdd8819efe4d535b72e19f7c0 100644
--- a/extensions/browser/BUILD.gn
+++ b/extensions/browser/BUILD.gn
@@ -41,6 +41,9 @@ source_set("browser_context_keyed_service_factories") {
 jumbo_source_set("browser_sources") {
   visibility = [ "./*" ]
 
+  # //brave/components/greaselion/browser uses this.
+  visibility += [ "//brave/components/greaselion/browser", ]
+
   sources = [
     "activity.cc",
     "activity.h",
@@ -406,6 +409,7 @@ jumbo_source_set("browser_sources") {
     "//third_party/blink/public/mojom/frame",
     "//ui/display",
   ]
+  if (brave_chromium_build) { deps += [ "//brave/extensions/browser"] }
 
   public_deps = [
     "//content/public/common",
