diff --git a/third_party/blink/renderer/core/page/create_window.cc b/third_party/blink/renderer/core/page/create_window.cc
index 9430b4a8120420188c7c8690cb684b4a534a210b..25626da03ac901140e92c93c6f1d714007963b18 100644
--- a/third_party/blink/renderer/core/page/create_window.cc
+++ b/third_party/blink/renderer/core/page/create_window.cc
@@ -306,7 +306,7 @@ Frame* CreateNewWindow(LocalFrame& opener_frame,
   // TODO(dmurph): Don't copy session storage when features.noopener is true:
   // https://html.spec.whatwg.org/C/#copy-session-storage
   // https://crbug.com/771959
-  CoreInitializer::GetInstance().CloneSessionStorage(old_page,
+  CoreInitializer::GetInstance().CloneSessionStorage(old_page, &opener_frame,
                                                      new_namespace_id);
 
   Page* page = old_page->GetChromeClient().CreateWindow(
